### ISSUE TITLE
segfault in Get_OneSNP_Geno

### ISSUE THREAD

**Issue by apoliakov** [weizhouUMICH/SAIGE#188]
Hey @weizhouUMICH ! Been a while :)

We're getting a new kind of crash in step 1 variance ratio estimation. It iterates over a few numbers of snps and then somehow it decides to look at the "NAth marker" and that crashes it:

```
91476 th marker
G0 0 0 0 0 0 0 0 0 0 0
iter from getPCG1ofSigmaAndVector 3
t1
t1again
t2-t1
   user  system elapsed
  0.004   0.000   0.004
95960 th marker
G0 0 0 0 0 0 0 0 0 0 0
iter from getPCG1ofSigmaAndVector 2
t1
t1again
t2-t1
   user  system elapsed
  0.002   0.000   0.002
NA th marker

 *** caught segfault ***
```

Don't know if this is another case of "we have an unbalanced file" or something else. Attaching the full [crash.log](https://github.com/weizhouUMICH/SAIGE/files/4631938/crash.log) .

SAIGE version:
```
> packageDescription('SAIGE')
Package: SAIGE
Type: Package
Title: Efficiently controlling for case-control imbalance and sample
        relatedness in single-variant assoc tests (SAIGE) and
        controlling for sample relatedness in region-based assoc tests
        in large cohorts and biobanks (SAIGE-GENE)
Version: 0.38
Date: 2020-05-04
Author: Wei Zhou, Seunggeun Lee, Cristen Willer
Maintainer: Wei Zhou <zhowei@umich.edu>
Description: an R package that implements the Scalable and Accurate
        Implementation of Generalized mixed model that uses the
        saddlepoint approximation (SPA)(mhof, J. P. , 1961; Kuonen, D.
        1999; Dey, R. et.al 2017) and large scale optimization
        techniques to calibrate case-control ratios in logistic mixed
        model score tests (Chen, H. et al. 2016) in large PheWAS.
        Estimated effect sizes are provided. Support VCF,BCF and SAV as
        input format for dosages using the SAVVY library and support
        BGEN as input format for dosages using the BGEN library. QR
        decomposition for the covariate matrix. Allow models with no
        covariates. Allow using large number of genetic markers to
        construct GRM (e.g. > 600,000 markers). Fixed a Bug for the
        Tstat output. Use coefficent of variations for trace estimators
        and variane ratio estimation. Fixed a bug for colSums() when
        there is no covariate. BETA and Tstat are now for the alt
        allele for both quantitative and binary traits. Added an option
        for leave-one-chromosome-out (LOCO), cut off for the
        coefficient of variation for trace estimates and variance ratio
        estimates. Update the savvy library. conditional analysis and
        gene-based test with sparse GRM. Add option for categorical
        variance ratio and sparseSigma for single-variant analysis for
        quantitattive traits. (bianry traits needs more work). Add step
        1 using GMMAT package. Work with R-3.5.1. Update eta for
        fitting null GLMM for binary traits. Add a script
        createSparseGRM.R to create a sparse GRM. Use sparse matrix for
        gene-based tests. Fixed a bug for step 1 (updating eta).
        Replace pcg with solve for sparseGRM. Remove markers with
        negative variance from gene-based tests. Speed up the step 1
        for quantitative traits. Use sparse GRM to estiamte the inital
        tau. Fix a bug for step 2 when phi is very small. Add the
        missing arguments cateVarRatioMinMACVecExclude and
        cateVarRatioMaxMACVecInclude to step 1 functions. Compared to
        0.35.5.6, this version has master branch merged in. The
        following comments are copied from master that are missed since
        the master-gene was branched out from master.  Update SAIGE as
        SPAtest library is updated (bug fixed). A bug is fixed in the
        savvy library for vcf and sav format input. A bug is fixed for
        step 1 for binary traits, which may not affect much on step 1
        results. Break when tau is 0 for quantitative traits in step 1.
        Add code to check sample size in sample file and dosage file.
        remove covariates in case of perfect separation. handle missing
        dosages. 0.35.6.3:add a function to extract the diag of GRM.
        0.35.7: clean code. 0.35.8: merge 0.35.7 to 0.29.7 in the
        master branch. 0.35.8.1: fix some errors in documentation and
        warning message for binary traits. 0.35.8.2: minor changes-fix
        error message, change MAC to MAF, add a line to check if the
        chomosome in plink file is numeric or not, add rsid to the
        header when input file is bgen. 0.35.8.3: fix a bug in the
        function getCovM_nopcg, which affected the conditional analysis
        for binary traits. Merge hyacz/master to use cget to manage
        superlu.0.35.8.5: implement the function to account for
        unbalanced case control ratio for gene-based tests. 0.35.8.6:
        fixed the output bug when the genotype matrix has rank 1 for
        binary phenotypes and add an argument minMAFtoConstructGRM for
        step 0 and step 1 to allow users to specify the minumum MAF of
        markers used to construct GRM. 0.35.8.7: fixed the bug when
        there is no covariate specified, added an argument
        IsOutputNinCaseCtrl for step 2 to allow for output sample sizes
        in cases and controls for binary traits in the output file, and
        fixed the out of boundary bug for LOCO. 0.35.8.8: Fixes a
        matrix inversion issue in the null model and adds an optional
        argument for the null computation to remove binary covariates
        with low counts by juhis. 0.36.1: 1. fixed the freq calculation
        for mean impute for missing genotypes in plinkFile 2. Diagonal
        elements of GRM are now estimated using markers in plinkFile
        with MAF >= minMAFforGRM 3. Conditional analysis for gene- or
        region-based test for binary traits is now accounting for
        case-control imbalance 4. plain dosage files are no longer
        supported for step 2 so no external boost_iostream library is
        needed. 0.36.2: The option weights.beta.common is not fully
        correctly developed, so we make weights.beta.common equal to
        weights.beta.rare for now. Instead of output NA for SKAT-O p
        values when the function SKAT:::Met_SKAT_Get_Pvalue failed,
        output 2*min(SKAT p, Burden p, 0.5). allow for specifying
        customized weights for markers in gene- or region-based tests
        with arguments in step 2: weightsIncludeinGroupFile and
        weights_for_G2_cond. 0.36.3: Add option
        IsOutputBETASEinBurdenTest in step 2 to output effect sizes for
        burden tests. 0.36.3.1: minor changes. 0.36.3.2: fixed a bug
        for gene-based conditioning tests with multipel conditioning
        markers and re-check markers after dropping samples with
        missing dosages/genotypes in gene-based tests. 0.36.3.3: trying
        to fix a bug when there is only one sample or marker left after
        dropping samples with missing dosages/genotypes in gene-based
        tests. 0.36.4: add an option includeNonautoMarkersforVarRatio
        in step 1. If TRUE, non-autosomal markers are also used for
        variance ratio estimation, which will make the algorithm more
        appropriate for assoc tests for non-autosomal markers; use the
        new function with sparse sigma for p-values for single variants
        in gene-based tests; assign AF to be 0 if all samples have
        missing genotypes or dosages. 0.36.4.1: Trying to fix a bug
        when minMAFforGRM is set and LOCO=TRUE. 0.36.4.2: fix a typo to
        extract p.value. 0.36.5: fix an issue for LOCO=TRUE. This issue
        was introduced when the option minMAFforGRM was introduced.
        0.36.5.1: add the option SPAcutoff, If the test statistic lies
        within the standard deviation cutoff of the mean, p-value based
        on traditional score test is returned. Otherwise, SPA will be
        applied. Default value of SPAcutoff is 2 (corresponding
        p.value.NA 0.05). 0.36.6. add an option
        IsOutputHetHomCountsinCaseCtrl to output the heterozygous and
        homozygous counts in cases and controls. 0.37. fixed a bug for
        AC values when bgen input with missing dosages was used and the
        missing doages are mean imputed. 0.38. further fixed the bug
        for output the allele 2 when bgen input with missing dosages
        was dropped and sampleFile is no longer needed if VCF file is
        used in Step 2.  add --IsOverwriteVarianceRatioFile in step 1
        to overwrite the variance ratio file
License: GPL (>= 2)
Imports: Rcpp (>= 0.12.12), RcppParallel, Matrix
LinkingTo: Rcpp, RcppArmadillo, RcppParallel, data.table, SPAtest (==
        3.0.2), RcppEigen, Matrix, methods, BH, optparse, SKAT,
        MetaSKAT
SystemRequirements: GNU make
RoxygenNote: 7.0.1
NeedsCompilation: yes
Encoding: UTF-8
Packaged: 2020-05-12 20:47:30 UTC; scidb
RemoteType: github
RemoteHost: api.github.com
RemoteRepo: saige
RemoteUsername: weizhouUMICH
RemoteRef: master
RemoteSha: 8f3021de7c0969ee5a6f29c2b157ebfc33e3a498
GithubRepo: saige
GithubUsername: weizhouUMICH
GithubRef: master
GithubSHA1: 8f3021de7c0969ee5a6f29c2b157ebfc33e3a498
Depends: R (>= 3.5.0)
Built: R 3.5.2; x86_64-redhat-linux-gnu; 2020-05-12 20:50:55 UTC; unix

-- File: /home/scidb/R/x86_64-redhat-linux-gnu-library/3.5/SAIGE/Meta/package.rds 
```

Let us know what you think! Thanks a lot!

**Comment by weizhouUMICH** [comment]
Hi @apoliakov, 

It seems the error is because that for variance ratio estimation for singletons, there are not enough singletons available in the Plink file. After 90 of them were used, the CV is still under the CV cutoff,  so the program asks for 10 more markers, but there are only 92 singletons in total in the plink file. Could you please either add more rare variants to the plink file or decrease the the sample relatedness coefficients cutoff in Step 0? 1. With the new plink file, you may skip the model fitting and re-estimate the variance ratio with --skipModelFitting=TRUE. 2. Note that the current CV of the variance ratio estimates is still much higher than the cutoff 0.001. I think decreasing the sample relatedness coefficients cutoff by including more samples in the sparse GRM, the variance ratio estimation can converge faster with a smaller subset of singletons. 

Thanks!
Wei

Thanks,
Wei

**Comment by apoliakov** [comment]
Thanks @weizhouUMICH !

I can confirm the thing worked with a different file. So we'll just use a larger file. Part of this is my fault - I was trying to see if I can use a smaller step 1 file to squeeze some performance gains out of it. Thanks for your help.

Step 1 can take a while. It can be 1 hour per phenotype. So if we're running 1000 new phenotypes, even on 200 machines, it is still 5 hours. Not trivial :) And sometimes we see a single Step 1 take 2 or 3 hours too. I thought I was being "clever" with smaller file but, looks like I did the opposite.

Thanks again.

**Comment by weizhouUMICH** [comment]
We have just released a new version 1.0.0. It has computational efficiency improvements for both Step 1 and Step 2 for single-variant and set-based tests. We have created a new program github page https://github.com/saigegit/SAIGE with the documentation provided https://saigegit.github.io/SAIGE-doc/
The program will be maintained by multiple SAIGE developers there. The docker image has been updated. Please feel free to try the version 1.0.0 and report issues if any.

Thanks!
Wei
